### PR TITLE
chore(userspace/engine): changed format of the 'output' field in the JSON payload

### LINKED ISSUE #2985: Change the format of the 'output' field in the JSON payload
**Motivation**

Right now, the default output for Falco is stdout with `basic text` as format. The generated log lines follow this pattern `<timestamp> <Priority> <output>`:

```
14:37:27.505989596: Warning Detected ptrace PTRACE_ATTACH attempt (proc_pcmdline=%proc.pcmdline evt_type=%evt.type user=%user.name user_uid=%user.uid user_loginuid=%user.loginuid process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty exe_flags=%evt.arg.flags %container.info)
```

By choosing the JSON format we get the same content in the `output` field, with same elements `timestamp` and `Priority`, but these elements are also contained in specific fields of the JSON:

```json
{
    "output": "14:37:27.505989596: Warning Detected ptrace PTRACE_ATTACH attempt (proc_pcmdline=%proc.pcmdline evt_type=%evt.type user=%user.name user_uid=%user.uid user_loginuid=%user.loginuid process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty exe_flags=%evt.arg.flags %container.info)",
    "priority": "WARNING",
    "rule": "PTRACE attached to process",
    "time": "2023-12-20T14:37:27.505989596Z",
    "output_fields": {
        "container.info": "container.info",
        "evt.arg.flags": "evt.arg.flags",
        "evt.type": "evt.type",
        "proc.cmdline": "proc.cmdline",
        "proc.exepath": "proc.exepath",
        "proc.name": "proc.name",
        "proc.pcmdline": "proc.pcmdline",
        "proc.pname": "proc.pname",
        "proc.tty": "proc.tty",
        "user.loginuid": "user.loginuid",
        "user.name": "user.name",
        "user.uid": "user.uid"
    },
    "hostname": "host-7.local",
    "source": "syscalls",
    "tags": [
        "maturity_stable",
        "host",
        "container",
        "process",
        "mitre_privilege_escalation",
        "T1055.008"
    ]
}
```

It implies we have a duplication of the information between the fields, and it creates a mess for systems trying to deduplicate alerts. See [this issue for Falcosidekick](https://github.com/falcosecurity/falcosidekick/issues/711).

**Feature**

I propose to allow an opt-in to remove the timestamp and/or the priority from the `outpout` when the JSON format is used.

**Alternatives**

I created a PR for Falcosidekick allowing to reformat the output field. See [PR#729](https://github.com/falcosecurity/falcosidekick/pull/729)

**Additional context**

### PR DESCRIPTION
**Pull request by h4l0gen**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:
'timestamp' and 'priority' format of the 'output' field in the JSON format has been removed. 
**Which issue(s) this PR fixes**:
Change the format of the 'output' field in the JSON payload

Fixes #2985 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
chore(userspace/engine): changed format of the 'output' field in the JSON payload
```

### CODE REVIEW

File: userspace/engine/formats.cpp [hunk 1]
```diff
@@ -44,7 +44,6 @@ std::string falco_formats::format_event(gen_event *evt, const std::string &rule,
 	formatter = m_falco_engine->create_formatter(source, format);
 
 	// Format the original output string, regardless of output format
```
Comment (Andreagit97) [inline userspace/engine/formats.cpp:46]: can we remove this comment? it seems we don't need it anymore

**Review by Andreagit97 — COMMENTED** [review]
Thank you for this!

Linking here an example of what this change means https://github.com/falcosecurity/falco/issues/2985#issuecomment-1898559054. Read the full issue to gather full context.

File: userspace/engine/formats.cpp (hunk 1 above)
Comment (h4l0gen) [reply userspace/engine/formats.cpp]: Sure @Andreagit97, I removed the comment and committed the changes. Is everything else fine? Please let me know if any more changes are required in PR 😀

**Review by leogr — COMMENTED** [review]
As per the [original proposal](https://github.com/falcosecurity/falco/issues/2985#issuecomment-1898559054), this behavior must be opt-in for users.

So we need to add a new config key in `falco.yaml` to enable this behavior. The previous format must be preserved when this new setting is off (default).

**Comment by h4l0gen** [comment]
@leogr here is the discussion about this https://kubernetes.slack.com/archives/CMWH3EH32/p1706275031447379?thread_ts=1705577834.276719&cid=CMWH3EH32 @Andreagit97 proposed that we don't really need current output format and  removing timestamp tag permanently would be good for user experience, WDYT about it @leogr 🤔

**Comment by leogr** [comment]
> @leogr here is the discussion about this https://kubernetes.slack.com/archives/CMWH3EH32/p1706275031447379?thread_ts=1705577834.276719&cid=CMWH3EH32 @Andreagit97 proposed that we don't really need current output format and removing timestamp tag permanently would be good for user experience, WDYT about it @leogr 🤔

Sorry, I missed that discussion :angel:  So, ignore my previous comment for now. If we get a consensus on that, it will be ok for me, too. Just a note for maintainers: if we agree to change the JSON "output" field without an opt-in, this will introduce a (minor) breaking change that needs to be signaled in the release note.

**Comment by Andreagit97** [comment]
> this will introduce a (minor) breaking change that needs to be signaled in the release note.

Yes! Let's see what other maintainers think about that @falcosecurity/falco-maintainers

**Comment by FedeDP** [comment]
`testing` framework needs adaption.
Btw i think this is for 0.38.0, right?

> Yes! Let's see what other maintainers think about that @falcosecurity/falco-maintainers

I agree, it  should be signaled in the rn, and i'd add an entry in the 0.38.0 breaking changes; even if minor, it is a breaking change.

**Comment by Andreagit97** [comment]
if all maintainers agree, I think we can go with this :)

**Comment by leogr** [comment]
> if all maintainers agree, I think we can go with this :)

Ok, for me.

Does anyone disagree with introducing this small breaking change?
cc @incertum @LucaGuerra

**Comment by incertum** [comment]
LGTM!

**Comment by h4l0gen** [comment]
@FedeDP @leogr @Andreagit97 @incertum build is fixed😀. please take a look into it and provide your feedback or any changes if required.

**Comment by FedeDP** [comment]
Thanks!
Can you squash the commits together?

**Comment by Andreagit97** [comment]
i should have fixed it, let's see if the CI passes

**Comment by h4l0gen** [comment]
@FedeDP, I squashed all three commits. Sorry for the delay in fixing this build 😅, as I've also started learning and contributing to falcosecurity/event-generator for GSoC project idea.

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by Andreagit97** [comment]
oh since the format is changed we need to change all the associated tests :/ we will take care of that, don't worry!

**Comment by h4l0gen** [comment]
@Andreagit97 if output field format also documented somewhere , we need to take care of that too :)

**Comment by leogr** [comment]
image
@FedeDP :eye:

**Comment by FedeDP** [comment]
@leogr yes, we need to fix 3 tests on https://github.com/falcosecurity/testing framework:
```
  TestFalco_Legacy_StdoutOutputJsonStrict
  TestFalco_Legacy_NullOutputField
  TestFalco_Legacy_TimeIso8601
```

**Comment by Andreagit97** [comment]
> oh since the format is changed we need to change all the associated tests :/ we will take care of that, don't worry!

folks, i've already written this message, i will take care of these tests when i have time. Let's avoid to ping everyone in the discussion

**Comment by FedeDP** [comment]
Thanks Andre!

**Comment by Andreagit97** [comment]
uhm fixing tests i noticed that probably we are obtaining something that we don't want...
Before this PR the output of a simple rule was:

```json
{"hostname":"test-falco-hostname","output":"2016-08-04T16:17:57.881781397+0000: Warning An open was seen (command=cat /dev/null)","priority":"Warning","rule":"open_from_cat","source":"syscall","tags":["filesystem","process","testing"],"time":"2016-08-04T16:17:57.881781397Z", "output_fields": {"evt.time.iso8601":1470327477881781397,"proc.cmdline":"cat /dev/null"}}
```

While now the output is:

```json
{"hostname":"test-falco-hostname","output":"{\"evt.time.iso8601\":1470327477881781397,\"proc.cmdline\":\"cat /dev/null\"}","priority":"Warning","rule":"open_from_cat","source":"syscall","tags":["filesystem","process","testing"],"time":"2016-08-04T16:17:57.881781397Z", "output_fields": {"evt.time.iso8601":1470327477881781397,"proc.cmdline":"cat /dev/null"}}
```

As you can see the `output:` is changed, but what scares me is that we don't have anymore `(command=cat /dev/null)` but we have `"proc.cmdline\":\"cat /dev/null\"`, not sure this is what we want since this is a big breaking change because the names change

**Comment by incertum** [comment]
@Andreagit97 thanks for checking, wasn't aware we changed that to JSON. No this is not what we want and what is advertised as change in this PR.

**Comment by leogr** [comment]
> As you can see the `output:` is changed, but what scares me is that we don't have anymore `(command=cat /dev/null)` but we have `"proc.cmdline\":\"cat /dev/null\"`, not sure this is what we want since this is a big breaking change because we are completely changing the output format, not just removing the 3 fields `time`, `priority`, `rule`

Totally agree. This PR should not introduce such a breaking change.

**Comment by FedeDP** [comment]
Since there is no agreement here, i will move this to the next milestone.
/milestone 0.39.0

**Comment by leogr** [comment]
@lg @Issif can you take a look pls?

**Comment by FedeDP** [comment]
Moving to TBD milestone; feel free to ping me in case anybody is willing to work on this.
/milestone TBD

**Comment by leogr** [comment]
Superseded by https://github.com/falcosecurity/falco/pull/3314

/close

**Pull request closed without merging**